apm-agent-plugins/apm-error-logging-plugin/src/main/java/co/elastic/apm/agent/errorlogging/AbstractLoggerErrorCapturingInstrumentation.java

```diff
@@ -42,50 +42,37 @@
 
 public abstract class AbstractLoggerErrorCapturingInstrumentation extends TracerAwareInstrumentation {
 
-    @SuppressWarnings({"WeakerAccess"})
-    @VisibleForAdvice
-    public static final ThreadLocal<Boolean> nestedThreadLocal = new ThreadLocal<Boolean>() {
-        @Override
-        protected Boolean initialValue() {
-            return Boolean.FALSE;
-        }
-    };
-
     @Override
     public Class<?> getAdviceClass() {
         return LoggingAdvice.class;
     }
 
-    @Override
-    public boolean indyPlugin() {
-        return false;
-    }
-
     public static class LoggingAdvice {
 
-        @Advice.OnMethodEnter(suppress = Throwable.class)
-        public static void logEnter(@Advice.Argument(1) Throwable exception,
-                                    @Advice.Local("nested") boolean nested,
-                                    @Advice.Origin Class<?> clazz,
-                                    @Advice.Local("error") @Nullable ErrorCapture error) {
-            nested = nestedThreadLocal.get();
-            if (!nested) {
-                error = tracer.captureException(exception, tracer.getActive(), clazz.getClassLoader());
+        private static final CallDepth callDepth = CallDepth.get(LoggingAdvice.class);
+
+        @Nullable
+        @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
+        public static Object logEnter(@Advice.Argument(1) Throwable exception,
+                                        @Advice.Origin Class<?> clazz) {
+            if (!callDepth.isNestedCallAndIncrement()) {
+                ErrorCapture error = tracer.captureException(exception, tracer.getActive(), clazz.getClassLoader());
                 if (error != null) {
                     error.activate();
                 }
-                nestedThreadLocal.set(Boolean.TRUE);
+                return error;
             }
+            return null;
         }
 
-        @Advice.OnMethodExit(suppress = Throwable.class)
-        public static void logExit(@Advice.Local("nested") boolean nested,
-                                   @Advice.Local("error") @Nullable ErrorCapture error) {
-            if (error != null) {
-                error.deactivate().end();
+        @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class, inline = false)
+        public static void logExit(@Advice.Enter @Nullable Object errorCaptureObj) {
+            if (callDepth.isNestedCallAndDecrement()) {
+                return;
             }
```

Comment:
Can be simplified as there's already a `if (errorCaptureObj instanceof ErrorCapture)` check afterwards. That check is only true if the error object is not null.

```suggestion
            callDepth.decrement();
```

Reply:
simplified